### PR TITLE
Adding Ardupilot SITL to MAVSDK CI 

### PR DESCRIPTION
**Pull request by ykhedar**
Following discussion on Issue #728 for adding Ardupilot support in MAVSDK (related issues #1562 and #1568), this pull request adds Ardupilot SITL (Copter, Rover) along with necessary changes needed to run it successfully. Currently onnly 3/34 Integrations tests pass. Some notes about changes made:

1. Added Dockerfiles each for Copter and Rover Firmware versions 4.1.2
2. Edited the script for building and pushing the new docker files.
3. Edited `run-sitl-tests.sh`, `start_px4_sitl.sh` and `stop_px4_sitl.sh` to include ardupilot related sitl codes. May need to change the file names in the future. e.g. `start_px4_sitl.sh` -> `start_sitl.sh`.
4. Added code block for running ardupilot sitl into the github actions file.

Requesting Review and Suggestions,
ykhedar

P.S. Congrats on the v1.0 release of MAVSDK :)

### CODE REVIEW

File: tools/run-sitl-tests.sh [hunk 1]
```diff
@@ -7,9 +7,29 @@ if [ "$#" -ne 1 ]; then
     exit 1
 fi
 
-PX4_FIRMWARE_DIR=$1
+echo "PX4 Version is: " ${PX4_VERSION}
```
Comment (JonasVautherin) [inline tools/run-sitl-tests.sh:10]: Should that line exist? Feels like it's already done below (line 13), right?

**Review by JonasVautherin — COMMENTED** [review]
Nice! It requires the APM containers to be pushed on the mavsdk docker registry, right? I see that the CI is failing to pull them at the moment.

File: tools/run-sitl-tests.sh (hunk 1 above)
Comment (ykhedar) [reply tools/run-sitl-tests.sh]: my bad :) .Removed it in the latest commit.

**Comment by ykhedar** [comment]
> Nice! It requires the APM containers to be pushed on the mavsdk docker registry, right? I see that the CI is failing to pull them at the moment.

exactly. I did all the tests on my account with all images pushed to my docker hub repo. After everything was working, i quashed all the commits into one so as to keep the pull request clean.

File: tools/start_px4_sitl.sh [hunk 2]
```diff
@@ -25,61 +25,119 @@ if [[ $AUTOSTART_SITL != 1 ]]; then
     exit 0
 fi
 
-if [[ -n "$PX4_FIRMWARE_DIR" ]]; then
-    px4_firmware_dir=$PX4_FIRMWARE_DIR
+# PX4 specific SITL Run Code
```
Comment (julianoes) [inline tools/start_px4_sitl.sh:28]: I agree we should change the name of these scripts.

**Review by julianoes — COMMENTED** [review]
Looks quite clean, thank you @ykhedar. Do the tests actually succeed?

**Review by julianoes — COMMENTED** [review]
Looks quite clean, thank you @ykhedar. Do the tests actually succeed?

**Comment by julianoes** [comment]
I have built and pushed the docker images.

**Comment by ykhedar** [comment]
> Looks quite clean, thank you @ykhedar. Do the tests actually succeed?


No. The tests are going on forever on the CI so I stopped the workflow. When I checked the logs, this was due to bad telemetry health condition inside a while loop in one of the tests for VTOL. Here is the link to the code:

[action_transition_multicopter_fixedwing](https://github.com/mavlink/MAVSDK/blob/cfd9feaa2786bb3850813ed3ea0ab27041ee7742/src/integration_tests/action_transition_multicopter_fixedwing.cpp#L37)

Should we add maybe a break condition in this test instead of letting the while loop run forever? Maybe do the same for all such while loops in all the tests? Just make the test fail if the while condition is not met after 60 seconds or 120 seconds?

**Comment by JonasVautherin** [comment]
> Maybe do the same for all such while loops in all the tests?

Generally I think it would make sense if the tests did not hang forever :+1:.

This said, for the health, is it failing because of a difference with Ardupilot? Can we fix `health_all_ok()` easily?

**Comment by julianoes** [comment]
We should limit the tests to 45mins, like the PX4 SITL ones.

And if you want to go through all of them, feel free, but I think it's quite a bit of work.

Do the tests run through for you locally?

**Comment by ykhedar** [comment]
> We should limit the tests to 45mins, like the PX4 SITL ones.
> 
> And if you want to go through all of them, feel free, but I think it's quite a bit of work.
> 
> Do the tests run through for you locally?

How does PX4 SITL implement the limit of 45 minutes? Cant seem to find the setting for this. 

On my local machine i saw 3 tests pass (i swear :D) which now seems to be incorrect. My bad :( . 

Should I add a simple timer using std::chrono in the while loop to wait for 30 seconds or so before the loop exits with a failed test? Or do you suggest something better?

**Comment by JonasVautherin** [comment]
> How does PX4 SITL implement the limit of 45 minutes? Cant seem to find the setting for this.

I think @julianoes was referring to [this](https://github.com/mavlink/MAVSDK/blob/main/.github/workflows/main.yml#L634).

**Comment by julianoes** [comment]
@ykhedar it would be best to use this function for checks with timeout:
https://github.com/mavlink/MAVSDK/blob/6ad7dbd424a5be32c0d4f3e61590fa6c1de3cb83/src/integration_tests/integration_test_helper.h#L69-L84

**Review by julianoes — DISMISSED** [review]
Nice!

**Comment by ykhedar** [comment]
> Nice!

Just made some changes to the commits (squashed them into one) and your review was made stale by git. Also fetched upstream and merged the new commits into ykhedar/MAVSDK.

Apparently 2 out of 34 tests are passing.

**Review by JonasVautherin — DISMISSED** [review]

**Comment by ykhedar** [comment]
Thanks @JonasVautherin for your review. Do we need to get all the Ardupilot related Integration tests to succeed before merging this pull request?

**Comment by JonasVautherin** [comment]
We already use `--gtest_filter` in the [script running the test](https://github.com/mavlink/MAVSDK/blob/main/tools/run-sitl-tests.sh#L15). What about adding `PX4` in the tests that only pass with PX4, and filter them out in the Ardupilot runs? Enabling a test for Ardupilot would then just require us to remove "PX4" from the test name, so they could be incrementally activated. What do you think?

**Comment by ykhedar** [comment]
Sounds like a good idea to me. I would do the necessary changes. Something like this:

`TEST_F(SitlTest, ActionGoto)` -> `TEST_F(SitlTest, PX4ActionGoto)` ?

**Comment by ykhedar** [comment]
Okay, I did the changes but it seems now some checks related to macOS-framework and ios are failing for some reason and i cannot think of why. Any ideas? :)

**Comment by JonasVautherin** [comment]
Grmbl I don't get why the ios build sometimes does that :disappointed:

**Review by JonasVautherin — APPROVED** [review]

**Review by julianoes — APPROVED** [review]
Thanks a lot for your work @ykhedar!

**Pull request merged**